src/ripe/base.py

```diff
@@ -70,6 +70,11 @@ def login(self, username = None, password = None, admin = None, token = None):
         admin = admin or self.admin
         token = token or self.token
         if token: return self.login_pid(token = token)
+        appier.verify(
+            self.username and self.password,
+            "Username and password not provided",
+            403
+        )
         url = self.base_url + ("signin_admin" if admin else "signin")
         contents = self.post(

```

Comment:
Why are we not getting a 403 from this POST operation?

Reply:
Because `username` and `password` are not defined, so it's a 400